rir/src/interpreter/cache.h

```diff
@@ -1,56 +1,104 @@
 #ifndef RIR_INTERPRETER_CACHE_H
 #define RIR_INTERPRETER_CACHE_H
 
+#include "R/r.h"
 #include "instance.h"
 
 namespace rir {
 
-// must be a power of 2 for modulus using & CACHE_MASK to work
-#define MAX_CACHE_SIZE 128
-#define CACHE_MASK (MAX_CACHE_SIZE - 1)
+// makes max cache size fit into 64bytes (with size)
+#define MAX_CACHE_SIZE 63
 #define ACTIVE_BINDING_MASK (1 << 15)
 #define BINDING_LOCK_MASK (1 << 14)
 #define IS_ACTIVE_BINDING(b) ((b)->sxpinfo.gp & ACTIVE_BINDING_MASK)
 #define BINDING_IS_LOCKED(b) ((b)->sxpinfo.gp & BINDING_LOCK_MASK)
+#define FRAME_LOCK_MASK (1 << 14)
+#define FRAME_IS_LOCKED(e) (ENVFLAGS(e) & FRAME_LOCK_MASK)
 
 #ifdef CACHE_ON_R_STACK
 // TODO: Create a version with a cache on the R_stack instead of the C stack
 #else
-typedef struct {
-    SEXP loc;
-} BindingCache;
-typedef BindingCache Cache;
+typedef SEXP BindingCacheEntry;
+struct BindingCache {
+    size_t length;
+    BindingCacheEntry entry[];
+};
 
-static RIR_INLINE void clearCache(Cache* cache, size_t cacheSize) {
-    memset(cache, 0, sizeof(Cache) * cacheSize);
+static RIR_INLINE void clearCache(BindingCache* cache) {
+    memset(cache->entry, 0, sizeof(BindingCacheEntry) * cache->length);
 }
 
-static RIR_INLINE SEXP cachedGetBindingCell(SEXP env, Immediate poolIdx,
-                                            Immediate cacheIdx,
-                                            InterpreterInstance* ctx,
-                                            Cache* cache,
-                                            bool smallCache) {
-    return cache[cacheIdx].loc;
+static RIR_INLINE SEXP cachedGetBindingCell(Immediate cacheIdx,
+                                            BindingCache* cache) {
+    SLOWASSERT(cacheIdx < cache->length);
+    return cache->entry[cacheIdx];
 }
 
-static RIR_INLINE void cachedSetBindingCell(Immediate cacheIdx, Cache* cache,
-                                            bool smallCache, R_varloc_t loc) {
-    cache[cacheIdx].loc = loc.cell;
+static RIR_INLINE void
+cachedSetBindingCell(Immediate cacheIdx, BindingCache* cache, R_varloc_t loc) {
+    SLOWASSERT(cacheIdx < cache->length);
+    cache->entry[cacheIdx] = loc.cell;
+}
+
+static void rirDefineVarWrapper(SEXP symbol, SEXP value, SEXP rho) {
```

Comment:
Why do we need this wrapper?

Reply:
the main slowdown in storage is from the fact that you use INCREMENT_NAMED on things which don't need them

Reply:
My concern is, we were executing the `INCREMENT_NAMED` previously. I did not change the semantics of the variable set, did I?

Reply:
yes, so in the cached lookups we had case that would check if the thing that we are overriding is the same. in this case you do not need to increment.

with your changes all the non-caching implementations lost that trick and they started incrementing named counts like crazy.

you probably didn't notice since you set the cache limit so high, it would never trigger (also seen by the fact that both the normal compilar and pir optimizer had several bugs, when the cache size would be reached...)